### PR TITLE
Upgrade configuration

### LINKED ISSUE #16: YAML support
Users should be able to load a yaml config file

### PR DESCRIPTION
**Pull request by chrhck**
Upgraded CFG dict to a class, added options to overwrite config using a yaml file or a dictionary
Closes #16

### CODE REVIEW

File: scripts/generate_trials.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        (seed,mean_ns,mean_ns_null,trial_data)=create_trial_data_file(
+            ana=ana,
+            ncpu=args.ncpus,
+            rss=rss,
+            pathfilename=trials_outfile,
+            n_trials=args.n_trials,
+            mean_n_sig=args.mean_ns,
+            tl=tl)
+
+    print(tl)
```
Comment (HansN87) [inline scripts/generate_trials.py]: this entire script should be moved to skyllh.examples (potentially file name use_config.py)
script currently uses features that are not on master yet.

File: skyllh/core/config.py [hunk 2]
```diff
@@ -67,6 +69,53 @@
 }
 
 
+class CFGClass(dict):
+
+    """
+    This class holds the global config state
```
Comment (HansN87) [inline skyllh/core/config.py]: per-martin style code: add "." at end of doc-string

File: skyllh/core/config.py [hunk 3]
```diff
@@ -67,6 +69,53 @@
...
+    def __init__(self, *args, **kwargs) -> None:
+        if CFGClass._is_instantiated:
+            raise RuntimeError("Can only instantiate CFGClass once")
+
+        super().__init__(*args, **kwargs)
+        CFGClass._is_instantiated = True
+
+    def from_yaml(self, yaml_file: str) -> None:
+        """
+        Update config with yaml file
```
Comment (HansN87) [inline skyllh/core/config.py]: per-martin style code: add "." at end of doc-string

File: skyllh/core/config.py [hunk 4]
```diff
@@ -67,6 +69,53 @@
...
+            yaml_file: str
+                path to yaml file
+        """
+
+        yaml_config = yaml.load(open(yaml_file), Loader=yaml.SafeLoader)
+        self.update(yaml_config)
+
+    def from_dict(self, user_dict: Dict[Any, Any]) -> None:
+        """
+        Creates a config from dictionary
```
Comment (HansN87) [inline skyllh/core/config.py]: per-martin style code: add "." at end of doc-string

File: skyllh/core/config.py [hunk 5]
```diff
@@ -67,6 +69,53 @@
...
+        super().__init__(*args, **kwargs)
+        CFGClass._is_instantiated = True
+
+    def from_yaml(self, yaml_file: str) -> None:
+        """
+        Update config with yaml file
+
+        Parameters:
+            yaml_file: str
+                path to yaml file
```
Comment (HansN87) [inline skyllh/core/config.py]: per-martin style code: add "." at end of doc-string. Capitalize path.

File: skyllh/core/config.py [hunk 6]
```diff
@@ -67,6 +69,53 @@
...
+
+    def from_dict(self, user_dict: Dict[Any, Any]) -> None:
+        """
+        Creates a config from dictionary
+
+        Parameters:
+            user_dict: dict
+
+        Returns:
+            dict
```
Comment (HansN87) [inline skyllh/core/config.py]: What's in the dict that is returned? Can you add a short description?

File: skyllh/core/config.py [hunk 7]
```diff
@@ -6,13 +6,14 @@
 
 import os.path
 import sys
+from typing import Any, Dict, Iterator, KeysView, ItemsView, ValuesView
 
-from astropy import units
+from astropy import units  # type: ignore
```
[the remark this answers is not in the document]
Comment (chrhck) [reply skyllh/core/config.py]: yes, this prevents mypy from throwing an error due to missing type hints in astropy

File: examples/generate_trials.py [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+parser.add_argument("--n_trials", "-nt", help="number of trials per job", type=int, default=10)
+parser.add_argument("--mean_ns", "-ns", help="mean number of nsig injected per job", type=float, default=0.0)
+parser.add_argument("--ncpus", "-nc", help="number of CPUs to use", type=int, default=1)
+parser.add_argument("--rss_seed", "-rs", help="random_number_seed", type=int, default=5)
+parser.add_argument("--gseed", "-gs", help="gamma seed for minimizer", type=float, default=3.0)
+
+args=parser.parse_args()
+print(args)
+
+# DO NOT USE PHOTOSPLINE
```
Comment (HansN87) [inline examples/generate_trials.py]:
```suggestion
# DO NOT USE PHOTOSPLINE.
```

File: examples/generate_trials.py [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+skyllh.core.pdf.PHOTOSPLINE_LOADED = False
+
+CFG.from_yaml(args.config)
+
+# Define datasets (todo: generalize to other analyses)
```
Comment (HansN87) [inline examples/generate_trials.py]:
```suggestion
# Define datasets (todo: generalize to other analyses).
```

File: examples/generate_trials.py [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+datasets = dsc.get_datasets(ds_names)
+
+# Define the point source at TXS position.
+src_ra  = np.radians(CFG["analysis"]["default_source"]["ra"])
+src_dec = np.radians(CFG["analysis"]["default_source"]["dec"])
+source = PointLikeSource(src_ra, src_dec)
+
+tl = TimeLord()
+
+# Create analysis
```
Comment (HansN87) [inline examples/generate_trials.py]:
```suggestion
# Create analysis.
```

File: skyllh/core/config.py [hunk 11]
```diff
@@ -67,6 +69,51 @@
...
+class CFGClass(dict):
+
+    """
+    This class holds the global config state.
+
+    The class behaves like a dict, delegating all methods of the dict
+    interface to the underlying config dictionary.
+    """
+
+    # Keep track of whether this class has been instantiated
```
Comment (HansN87) [inline skyllh/core/config.py]:
```suggestion
    # Keep track of whether this class has been instantiated.
```

File: .github/workflows/pythonpackage.yml [hunk 12]
```diff
@@ -5,7 +5,7 @@ name: Python package
 
 on:
   push:
-    branches: [ master ]
+    branches: [ master, userconfig ]
```
Comment (martwo) [inline .github/workflows/pythonpackage.yml]: Why did you add the useconfig branch to the build bot workflow file? Once I merge this PR, the userconfig branch will be deleted automatically. For me this looks like an error.

File: examples/generate_trials.py [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+args=parser.parse_args()
+print(args)
+
+# DO NOT USE PHOTOSPLINE.
+skyllh.core.pdf.PHOTOSPLINE_LOADED = False
+
+CFG.from_yaml(args.config)
+
+# Define datasets (todo: generalize to other analyses).
+bp = CFG["paths"]["base_path"]
```
Comment (martwo) [inline examples/generate_trials.py]: I already used the "repository" section for the "base_path" setting. See also the i3skyllh/datasets/repository.py file

File: examples/generate_trials.py [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+dsc_name = CFG["analysis"]["name"]
```
Comment (martwo) [inline examples/generate_trials.py]: I'm not sure if "analysis"/"name" is the right address for the definition of the dataset collection. It is indeed analysis specific, so I suggest "analysis"/"dsc_name"

File: examples/baseconfig.yaml [hunk 15]
```diff
@@ -0,0 +1,12 @@
+paths:
+  base_path: /data/user/mwolf/data/
+  output_dir: /data/user/chaack/skyllh
+analysis:
+  name: NorthernTracks_v003p00_KDE_PDF_v003
+  datasets: IC79 IC86 MC
+  default_source:
+    ra: 77.358
+    dec: 5.693
+  livetime: 10
```
Comment (martwo) [inline examples/baseconfig.yaml]: Is it somehow possible to specify also the units in the yaml file? 10 could be 10days or 10years. The default unit in skyllh for livetime is days (which actually comes from skylab)

File: skyllh/core/config.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (martwo) [reply skyllh/core/config.py]: Maybe add this explanation as comment before the astropy import

File: examples/baseconfig.yaml (hunk 15 above)
Comment (chrhck) [reply examples/baseconfig.yaml]: in principle oyu can also use non simple types in yaml files, however I would rather add another option fo the unit, eg. `livetime_unit: "days"`

File: .github/workflows/pythonpackage.yml (hunk 12 above)
Comment (chrhck) [reply .github/workflows/pythonpackage.yml]: jup, this was just for testing

File: examples/baseconfig.yaml (hunk 15 above)
Comment (chrhck) [reply examples/baseconfig.yaml]: actually, I'll just use the internal_units["time"] in the script

Comment (martwo) [reply examples/baseconfig.yaml]: The internal unit for time is seconds. I would rather like the livetime_unit config tag. It would be nice if you could automatically map the specified unit (as str) to the astropy unit. The one can specify whatever unit which astropy supports
